packages/api/lambdas/create-reconciliation-report.js

```diff
@@ -0,0 +1,166 @@
+'use strict';
+
+const moment = require('moment');
+const {
+  aws: {
+    buildS3Uri,
+    DynamoDbScanQueue,
+    S3ListObjectsV2Queue,
+    s3
+  }
+} = require('@cumulus/common');
+
+/**
+ * Verify that all objects in an S3 bucket contain corresponding entries in
+ * DynamoDB, and that there are no extras in either S3 or DynamoDB
+ *
+ * @param {string} Bucket - the bucket containing files to be reconciled
+ * @param {string} filesTableName - the name of the files table in DynamoDB
+ * @returns {Promise<Object>} a report
+ */
+async function createReconciliationReportForBucket(Bucket, filesTableName) {
+  const s3ObjectsQueue = new S3ListObjectsV2Queue({ Bucket });
+  const dynamoDbFilesLister = new DynamoDbScanQueue({
+    TableName: filesTableName,
+    ExpressionAttributeNames: { '#b': 'bucket' },
+    ExpressionAttributeValues: { ':bucket': { S: Bucket } },
+    FilterExpression: '#b = :bucket'
+  });
+
+  let okFileCount = 0;
+  const onlyInS3 = [];
+  const onlyInDynamoDb = [];
+
+  let [nextS3Object, nextDynamoDbItem] = await Promise.all([s3ObjectsQueue.peek(), dynamoDbFilesLister.peek()]); // eslint-disable-line max-len
+  while (nextS3Object && nextDynamoDbItem) {
+    const nextS3Uri = buildS3Uri(Bucket, nextS3Object.Key);
+    const nextDynamoDbUri = buildS3Uri(Bucket, nextDynamoDbItem.key.S);
+
+    if (nextS3Uri < nextDynamoDbUri) {
```

Comment:
Glad you found a way to do this without needing to have millions-long lists locally.